akka-remote/src/main/scala/akka/remote/artery/Association.scala

```diff
@@ -198,32 +198,39 @@ private[remote] class Association(
             transport.system.deadLetters ! outboundEnvelope2
         case _ ⇒
           val outboundEnvelope = createOutboundEnvelope()
-          val offerOk =
-            if (transport.largeMessageChannelEnabled && isLargeMessageDestination(recipient))
-              largeQueue.offer(outboundEnvelope)
-            else
-              queue.offer(outboundEnvelope)
+          val queue = selectQueue(recipient)
+          val offerOk = queue.offer(outboundEnvelope)
           if (!offerOk)
             transport.system.deadLetters ! outboundEnvelope
       }
     } else if (log.isDebugEnabled)
       log.debug("Dropping message to quarantined system {}", remoteAddress)
   }
 
-  private def isLargeMessageDestination(recipient: OptionVal[RemoteActorRef]): Boolean = {
+  @tailrec
+  private def selectQueue(recipient: OptionVal[RemoteActorRef]): ProducerApi[OutboundEnvelope] = {
     recipient match {
       case OptionVal.Some(r) ⇒
-        if (r.cachedLargeMessageDestinationFlag ne null)
-          r.cachedLargeMessageDestinationFlag eq LargeDestination
-        else if (largeMessageDestinations.find(r.path.elements).isEmpty) {
-          r.cachedLargeMessageDestinationFlag = RegularDestination
-          false
-        } else {
-          log.debug("Using large message stream for {}", r.path)
-          r.cachedLargeMessageDestinationFlag = LargeDestination
-          true
+        r.cachedMessageDestinationFlag match {
+          case RegularDestination  ⇒ queue
+          case PriorityDestination ⇒ controlQueue
+          case LargeDestination    ⇒ largeQueue
+          case null ⇒
+            // only happens when messages are sent to new remote destination
+            // and is then cached on the RemoteActorRef
+            val elements = r.path.elements
+            if (priorityMessageDestinations.find(elements).isDefined) {
```

Comment:
Hm, can the WildcardIndex be made something like a Map, so after searching it you actually get back the Queue? (Or an index to the Array of Queues if the Queue instances are changing over time)